search/search-domain/src/main/java/io/camunda/search/filter/ProcessInstanceFilter.java

```diff
@@ -10,167 +10,229 @@
 import static io.camunda.util.CollectionUtil.*;
 
 import io.camunda.util.ObjectBuilder;
+import java.time.OffsetDateTime;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
-import java.util.function.Function;
+import java.util.stream.Collectors;
 
 public record ProcessInstanceFilter(
-    List<Long> processInstanceKeys,
-    List<String> processDefinitionIds,
-    List<String> processDefinitionNames,
-    List<Integer> processDefinitionVersions,
-    List<String> processDefinitionVersionTags,
-    List<Long> processDefinitionKeys,
-    List<Long> parentProcessInstanceKeys,
-    List<Long> parentFlowNodeInstanceKeys,
-    List<String> treePaths,
-    DateValueFilter startDate,
-    DateValueFilter endDate,
-    List<String> states,
+    List<Operation<Long>> processInstanceKeys,
+    List<Operation<String>> processDefinitionIds,
+    List<Operation<String>> processDefinitionNames,
+    List<Operation<Integer>> processDefinitionVersions,
+    List<Operation<String>> processDefinitionVersionTags,
+    List<Operation<Long>> processDefinitionKeys,
+    List<Operation<Long>> parentProcessInstanceKeys,
+    List<Operation<Long>> parentFlowNodeInstanceKeys,
+    List<Operation<String>> treePaths,
+    List<Operation<OffsetDateTime>> startDate,
+    List<Operation<OffsetDateTime>> endDate,
+    List<Operation<String>> states,
     Boolean hasIncident,
-    List<String> tenantIds)
+    List<Operation<String>> tenantIds)
     implements FilterBase {
 
   public static final class Builder implements ObjectBuilder<ProcessInstanceFilter> {
 
-    private List<Long> processInstanceKeys;
-    private List<String> processDefinitionIds;
-    private List<String> processDefinitionNames;
-    private List<Integer> processDefinitionVersions;
-    private List<String> processDefinitionVersionTags;
-    private List<Long> processDefinitionKeys;
-    private List<Long> parentProcessInstanceKeys;
-    private List<Long> parentFlowNodeInstanceKeys;
-    private List<String> treePaths;
-    private DateValueFilter startDate;
-    private DateValueFilter endDate;
-    private List<String> states;
+    private List<Operation<Long>> processInstanceKeys;
+    private List<Operation<String>> processDefinitionIds;
+    private List<Operation<String>> processDefinitionNames;
+    private List<Operation<Integer>> processDefinitionVersions;
+    private List<Operation<String>> processDefinitionVersionTags;
+    private List<Operation<Long>> processDefinitionKeys;
+    private List<Operation<Long>> parentProcessInstanceKeys;
+    private List<Operation<Long>> parentFlowNodeInstanceKeys;
+    private List<Operation<String>> treePaths;
+    private List<Operation<OffsetDateTime>> startDate;
+    private List<Operation<OffsetDateTime>> endDate;
+    private List<Operation<String>> states;
```

Comment:
💭 How about we use attribute names that describe what we store? Those are not `processInstanceKeys` anymore, those are `processInstanceKeyOperations`. 

That could help differentiate the method names as well: You can still have `processInstanceKeys(Long... keys)` that map to a list of operations, invoking `processInstanceKeyOperations(List<Operation<Long>> operations`.
That way, it is clearer where values and where operations are handed into methods.